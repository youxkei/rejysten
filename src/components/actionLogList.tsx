import type { ActionLogDocument } from "@/services/rxdb/collections/actionLog";

import { For } from "solid-js";

import { useRxDBService } from "@/services/rxdb";
import { createSubscribeAllSignal } from "@/services/rxdb/subscribe";

export function ActionLogList() {
  const { collections } = useRxDBService();

  const actionLogs$ = createSubscribeAllSignal(() =>
    collections.actionLogs.find({
      selector: { beginAt: { $gt: 0 } },
      sort: [{ beginAt: "asc" }],
    })
  );

  const tentativeActionLogs$ = createSubscribeAllSignal(() => collections.actionLogs.find({ selector: { beginAt: 0 } }));

  return (
    <>
      <For each={actionLogs$()}>{(actionLog) => <ActionLog actionLog={actionLog} />}</For>
      <For each={tentativeActionLogs$()}>{(actionLog) => <ActionLog actionLog={actionLog} />}</For>
    </>
  );
}

function ActionLog(props: { actionLog: ActionLogDocument }) {
  return (
    <div>
      <span>{props.actionLog.beginAt}</span>
      <span>→</span>
      <span>{props.actionLog.endAt}</span>
      <span>：</span>
      <span>{props.actionLog.text}</span>
    </div>
  );
}
