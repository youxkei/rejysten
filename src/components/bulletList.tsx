import type { JSX } from "solid-js";

import { styles } from "@/styles.css";

export function BulletList(props: { bullet: JSX.Element; item: JSX.Element; child: JSX.Element; isSelected: boolean }) {
  return (
    <div class={styles.bulletList.container}>
      <div class={styles.bulletList.bullet}>{props.bullet}</div>
      <div classList={{ [styles.bulletList.item]: true, [styles.selected]: props.isSelected }}>{props.item}</div>
      <div class={styles.bulletList.child}>{props.child}</div>
    </div>
  );
}
