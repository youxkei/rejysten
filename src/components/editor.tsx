import { createEffect, onCleanup } from "solid-js";

import { useEventService } from "@/services/event";
import { createSignalWithLock, useLockService } from "@/services/lock";
import { useStoreService } from "@/services/store";
import { Input } from "@/solid/input";
import { styles } from "@/styles.css";

export function Editor(props: { text: string }) {
  let input!: HTMLInputElement;

  const lock = useLockService();
  const { store } = useStoreService();
  const { emitEvent } = useEventService();

  const cursorPosition$ = createSignalWithLock(lock, () => store.editor.cursorPosition, -1, true);
  const text$ = createSignalWithLock(lock, () => props.text, "", true);

  createEffect(() => {
    input.focus();

    const pos = cursorPosition$();

    if (pos === -1) {
      const len = props.text.length;
      input.setSelectionRange(len, len);
    } else {
      input.setSelectionRange(pos, pos);
    }
  });

  let active = true;
  onCleanup(() => {
    active = false;
  });

  const onBlur = () => {
    if (active) {
      emitEvent({ pane: store.currentPane, mode: "insert", type: "leaveInsertMode" });
    }
  };

  return (
    <Input
      class={styles.editor}
      ref={input}
      onInput={(event) => emitEvent({ pane: store.currentPane, mode: "insert", type: "changeEditorText", newText: event.target.value })}
      onBlur={onBlur}
      value={text$()}
    />
  );
}
