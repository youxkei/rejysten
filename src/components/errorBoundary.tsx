import type { JSX } from "solid-js";

import { ErrorBoundary as SolidErrorBoundary } from "solid-js";

async function resetDB() {
  for (const db of await window.indexedDB.databases()) {
    window.indexedDB.deleteDatabase(db.name!);
  }
}

export function ErrorBoundary(props: { children: JSX.Element }) {
  return (
    <SolidErrorBoundary
      fallback={(err: { stack: string }, reset) => (
        <>
          <p>Something went wrong.</p>
          <pre>{err.stack}</pre>
          <button onClick={reset}>reset</button>
          <button onClick={resetDB}>reset DB</button>
        </>
      )}
    >
      {props.children}
    </SolidErrorBoundary>
  );
}
