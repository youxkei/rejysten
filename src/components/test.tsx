import type { JSXElement } from "solid-js";

import { Suspense } from "solid-js";

import { RxDBServiceProvider } from "@/services/rxdb";
import { createDatabaseCreatorForTest } from "@/services/rxdb/test";

export function TestWithRxDBService(props: { tid: string; children: JSXElement }) {
  return (
    <Suspense fallback={props.tid}>
      <RxDBServiceProvider databaseCreator={createDatabaseCreatorForTest(props.tid)}>
        {props.children}
      </RxDBServiceProvider>
    </Suspense>
  );
}

export { createCollectionsForTest } from "@/services/rxdb/test";
