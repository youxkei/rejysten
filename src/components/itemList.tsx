import { Show, For, createMemo, createEffect } from "solid-js";

import { useCollections, useSubscribe, useSubscribeAll } from "@/rxdb";
import { BulletList } from "@/components/bulletList";

export function ItemList(props: { id: string }) {
  return (
    <BulletList
      bullet={"â€¢"}
      item={<ItemListItem id={props.id} />}
      child={<ItemListChildren parentId={props.id} />}
    />
  );
}

export function ItemListItem(props: { id: string }) {
  const collections = useCollections();
  const listItem = useSubscribe(() =>
    collections()?.listItems.findOne(props.id)
  );

  return (
    <Show when={listItem()}>
      <span>{listItem()!.text}</span>
    </Show>
  );
}

export function ItemListChildren(props: { parentId: string }) {
  const collections = useCollections();
  const children = useSubscribeAll(() =>
    collections()?.listItems.find({ selector: { parentId: props.parentId } })
  );

  const sortedChildren = createMemo(
    () => {
      type ListItem = ReturnType<typeof children>[0];

      const unsortedChildren = children();
      if (unsortedChildren.length === 0) {
        return { inconsistent: false, children: unsortedChildren };
      }

      const sortedChildren = [] as ListItem[];
      const childMap = new Map<string, ListItem>();
      let currentChildId: string | null = null;

      for (const child of unsortedChildren) {
        if (child.prevId === "") {
          currentChildId = child.id;
        }
        childMap.set(child.id, child);
      }

      if (currentChildId === null) {
        return { inconsistent: true, children: [] };
      }

      while (currentChildId !== "") {
        const currentChild = childMap.get(currentChildId);

        if (currentChild === undefined) {
          return { inconsistent: true, children: [] };
        }

        sortedChildren.push(currentChild);
        currentChildId = currentChild.nextId;
      }

      if (unsortedChildren.length != sortedChildren.length) {
        return { inconsistent: true, children: [] };
      }

      return { inconsistent: false, children: sortedChildren };
    },
    { inconsistent: false, children: [] },
    { equals: (_, next) => next.inconsistent }
  );

  return (
    <>
      <For each={sortedChildren().children}>
        {(child) => <ItemList id={child.id} />}
      </For>
    </>
  );
}

import {
  render,
  waitForElementToBeRemoved,
  queryByText,
  findByText,
} from "solid-testing-library";

import { TestWithRxDB, createCollections } from "@/rxdb/test";

if (import.meta.vitest) {
  describe.each([
    {
      name: "swapped",
      listItems: [
        {
          id: "001",
          text: "root",
          prevId: "",
          nextId: "",
          parentId: "",
        },
        {
          id: "002",
          text: "foo",
          prevId: "003",
          nextId: "",
          parentId: "001",
        },
        {
          id: "003",
          text: "bar",
          prevId: "",
          nextId: "002",
          parentId: "001",
        },
      ],
    },
    {
      name: "nested",
      listItems: [
        {
          id: "001",
          text: "root",
          prevId: "",
          nextId: "",
          parentId: "",
        },
        {
          id: "002",
          text: "foo",
          prevId: "",
          nextId: "003",
          parentId: "001",
        },
        {
          id: "0021",
          text: "foofoo",
          prevId: "",
          nextId: "",
          parentId: "002",
        },
        {
          id: "003",
          text: "bar",
          prevId: "002",
          nextId: "",
          parentId: "001",
        },
      ],
    },
  ])("$name", ({ listItems }) => {
    test("text changes", async (ctx) => {
      const tid = ctx.meta.id;
      let collections = await createCollections(tid);
      await collections.listItems.bulkUpsert(listItems);

      const { container, unmount } = render(() => (
        <TestWithRxDB tid={tid}>
          <ItemList id="001" />
        </TestWithRxDB>
      ));

      await waitForElementToBeRemoved(() => queryByText(container, tid));
      ctx.expect(container).toMatchSnapshot();

      await (await collections.listItems.findOne("001").exec())!.atomicPatch({
        text: "changed root",
      });

      await findByText(container, "changed root");
      ctx.expect(container).toMatchSnapshot();

      await (await collections.listItems.findOne("002").exec())!.atomicPatch({
        text: "changed foo",
      });

      await findByText(container, "changed foo");
      ctx.expect(container).toMatchSnapshot();

      unmount();
    });
  });

  test("add child", async (ctx) => {
    const tid = ctx.meta.id;
    let collections = await createCollections(tid);
    await collections.listItems.bulkUpsert([
      {
        id: "001",
        text: "root",
        prevId: "",
        nextId: "",
        parentId: "",
      },
      {
        id: "002",
        text: "foo",
        prevId: "",
        nextId: "003",
        parentId: "001",
      },
      {
        id: "003",
        text: "bar",
        prevId: "002",
        nextId: "",
        parentId: "001",
      },
    ]);

    const { container, unmount } = render(() => (
      <TestWithRxDB tid={tid}>
        <ItemList id="001" />
      </TestWithRxDB>
    ));

    await waitForElementToBeRemoved(() => queryByText(container, tid));
    ctx.expect(container).toMatchSnapshot();

    await collections.listItems.bulkUpsert([
      {
        id: "003",
        text: "bar",
        prevId: "002",
        nextId: "004",
        parentId: "001",
      },
      {
        id: "004",
        text: "baz",
        prevId: "003",
        nextId: "",
        parentId: "001",
      },
    ]);

    await findByText(container, "baz");
    ctx.expect(container).toMatchSnapshot();

    unmount();
  });
}
