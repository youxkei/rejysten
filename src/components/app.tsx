import type { ChangeEvent } from "react";

import { useSelector, useDispatch } from "@/store";
import { app } from "@/slices/app";
import { useRxSync } from "@/rxdb";
import { RxdbSyncConfig } from "@/components/rxdbSyncConfig";
import { ItemList } from "@/components/itemList";

export function App() {
  useRxSync();

  const id = useSelector((state) => state.app.id);
  const dispatch = useDispatch();

  const onChange = (event: ChangeEvent<HTMLInputElement>) => {
    dispatch(app.actions.updateId({ id: event.target.value }));
  };

  return (
    <>
      {id === "" ? null : <ItemList id={id} />}
      <input value={id} onChange={onChange} />
      <RxdbSyncConfig />
    </>
  );
}
