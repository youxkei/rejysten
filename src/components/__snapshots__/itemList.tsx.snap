// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`'nested items' > text changes 1`] = `
<div>
  <div
    class="styles_styles_bulletList_container__1am92eo0"
  >
    <div
      class="styles_styles_bulletList_bullet__1am92eo1"
    >
      •
    </div>
    <div
      class="styles_styles_bulletList_item__1am92eo2"
    >
      <span>
        root
      </span>
    </div>
    <div
      class="styles_styles_bulletList_child__1am92eo3"
    >
      <div
        class="styles_styles_bulletList_container__1am92eo0"
      >
        <div
          class="styles_styles_bulletList_bullet__1am92eo1"
        >
          •
        </div>
        <div
          class="styles_styles_bulletList_item__1am92eo2"
        >
          <span>
            foo
          </span>
        </div>
        <div
          class="styles_styles_bulletList_child__1am92eo3"
        >
          <div
            class="styles_styles_bulletList_container__1am92eo0"
          >
            <div
              class="styles_styles_bulletList_bullet__1am92eo1"
            >
              •
            </div>
            <div
              class="styles_styles_bulletList_item__1am92eo2"
            >
              <span>
                foofoo
              </span>
            </div>
            <div
              class="styles_styles_bulletList_child__1am92eo3"
            />
          </div>
        </div>
      </div>
      <div
        class="styles_styles_bulletList_container__1am92eo0"
      >
        <div
          class="styles_styles_bulletList_bullet__1am92eo1"
        >
          •
        </div>
        <div
          class="styles_styles_bulletList_item__1am92eo2"
        >
          <span>
            bar
          </span>
        </div>
        <div
          class="styles_styles_bulletList_child__1am92eo3"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`'nested items' > text changes 2`] = `
<div>
  <div
    class="styles_styles_bulletList_container__1am92eo0"
  >
    <div
      class="styles_styles_bulletList_bullet__1am92eo1"
    >
      •
    </div>
    <div
      class="styles_styles_bulletList_item__1am92eo2"
    >
      <span>
        changed root
      </span>
    </div>
    <div
      class="styles_styles_bulletList_child__1am92eo3"
    >
      <div
        class="styles_styles_bulletList_container__1am92eo0"
      >
        <div
          class="styles_styles_bulletList_bullet__1am92eo1"
        >
          •
        </div>
        <div
          class="styles_styles_bulletList_item__1am92eo2"
        >
          <span>
            foo
          </span>
        </div>
        <div
          class="styles_styles_bulletList_child__1am92eo3"
        >
          <div
            class="styles_styles_bulletList_container__1am92eo0"
          >
            <div
              class="styles_styles_bulletList_bullet__1am92eo1"
            >
              •
            </div>
            <div
              class="styles_styles_bulletList_item__1am92eo2"
            >
              <span>
                foofoo
              </span>
            </div>
            <div
              class="styles_styles_bulletList_child__1am92eo3"
            />
          </div>
        </div>
      </div>
      <div
        class="styles_styles_bulletList_container__1am92eo0"
      >
        <div
          class="styles_styles_bulletList_bullet__1am92eo1"
        >
          •
        </div>
        <div
          class="styles_styles_bulletList_item__1am92eo2"
        >
          <span>
            bar
          </span>
        </div>
        <div
          class="styles_styles_bulletList_child__1am92eo3"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`'nested items' > text changes 3`] = `
<div>
  <div
    class="styles_styles_bulletList_container__1am92eo0"
  >
    <div
      class="styles_styles_bulletList_bullet__1am92eo1"
    >
      •
    </div>
    <div
      class="styles_styles_bulletList_item__1am92eo2"
    >
      <span>
        changed root
      </span>
    </div>
    <div
      class="styles_styles_bulletList_child__1am92eo3"
    >
      <div
        class="styles_styles_bulletList_container__1am92eo0"
      >
        <div
          class="styles_styles_bulletList_bullet__1am92eo1"
        >
          •
        </div>
        <div
          class="styles_styles_bulletList_item__1am92eo2"
        >
          <span>
            changed foo
          </span>
        </div>
        <div
          class="styles_styles_bulletList_child__1am92eo3"
        >
          <div
            class="styles_styles_bulletList_container__1am92eo0"
          >
            <div
              class="styles_styles_bulletList_bullet__1am92eo1"
            >
              •
            </div>
            <div
              class="styles_styles_bulletList_item__1am92eo2"
            >
              <span>
                foofoo
              </span>
            </div>
            <div
              class="styles_styles_bulletList_child__1am92eo3"
            />
          </div>
        </div>
      </div>
      <div
        class="styles_styles_bulletList_container__1am92eo0"
      >
        <div
          class="styles_styles_bulletList_bullet__1am92eo1"
        >
          •
        </div>
        <div
          class="styles_styles_bulletList_item__1am92eo2"
        >
          <span>
            bar
          </span>
        </div>
        <div
          class="styles_styles_bulletList_child__1am92eo3"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`'swapped items' > text changes 1`] = `
<div>
  <div
    class="styles_styles_bulletList_container__1am92eo0"
  >
    <div
      class="styles_styles_bulletList_bullet__1am92eo1"
    >
      •
    </div>
    <div
      class="styles_styles_bulletList_item__1am92eo2"
    >
      <span>
        root
      </span>
    </div>
    <div
      class="styles_styles_bulletList_child__1am92eo3"
    >
      <div
        class="styles_styles_bulletList_container__1am92eo0"
      >
        <div
          class="styles_styles_bulletList_bullet__1am92eo1"
        >
          •
        </div>
        <div
          class="styles_styles_bulletList_item__1am92eo2"
        >
          <span>
            bar
          </span>
        </div>
        <div
          class="styles_styles_bulletList_child__1am92eo3"
        />
      </div>
      <div
        class="styles_styles_bulletList_container__1am92eo0"
      >
        <div
          class="styles_styles_bulletList_bullet__1am92eo1"
        >
          •
        </div>
        <div
          class="styles_styles_bulletList_item__1am92eo2"
        >
          <span>
            foo
          </span>
        </div>
        <div
          class="styles_styles_bulletList_child__1am92eo3"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`'swapped items' > text changes 2`] = `
<div>
  <div
    class="styles_styles_bulletList_container__1am92eo0"
  >
    <div
      class="styles_styles_bulletList_bullet__1am92eo1"
    >
      •
    </div>
    <div
      class="styles_styles_bulletList_item__1am92eo2"
    >
      <span>
        changed root
      </span>
    </div>
    <div
      class="styles_styles_bulletList_child__1am92eo3"
    >
      <div
        class="styles_styles_bulletList_container__1am92eo0"
      >
        <div
          class="styles_styles_bulletList_bullet__1am92eo1"
        >
          •
        </div>
        <div
          class="styles_styles_bulletList_item__1am92eo2"
        >
          <span>
            bar
          </span>
        </div>
        <div
          class="styles_styles_bulletList_child__1am92eo3"
        />
      </div>
      <div
        class="styles_styles_bulletList_container__1am92eo0"
      >
        <div
          class="styles_styles_bulletList_bullet__1am92eo1"
        >
          •
        </div>
        <div
          class="styles_styles_bulletList_item__1am92eo2"
        >
          <span>
            foo
          </span>
        </div>
        <div
          class="styles_styles_bulletList_child__1am92eo3"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`'swapped items' > text changes 3`] = `
<div>
  <div
    class="styles_styles_bulletList_container__1am92eo0"
  >
    <div
      class="styles_styles_bulletList_bullet__1am92eo1"
    >
      •
    </div>
    <div
      class="styles_styles_bulletList_item__1am92eo2"
    >
      <span>
        changed root
      </span>
    </div>
    <div
      class="styles_styles_bulletList_child__1am92eo3"
    >
      <div
        class="styles_styles_bulletList_container__1am92eo0"
      >
        <div
          class="styles_styles_bulletList_bullet__1am92eo1"
        >
          •
        </div>
        <div
          class="styles_styles_bulletList_item__1am92eo2"
        >
          <span>
            bar
          </span>
        </div>
        <div
          class="styles_styles_bulletList_child__1am92eo3"
        />
      </div>
      <div
        class="styles_styles_bulletList_container__1am92eo0"
      >
        <div
          class="styles_styles_bulletList_bullet__1am92eo1"
        >
          •
        </div>
        <div
          class="styles_styles_bulletList_item__1am92eo2"
        >
          <span>
            changed foo
          </span>
        </div>
        <div
          class="styles_styles_bulletList_child__1am92eo3"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`add child 1`] = `
<div>
  <div
    class="styles_styles_bulletList_container__1am92eo0"
  >
    <div
      class="styles_styles_bulletList_bullet__1am92eo1"
    >
      •
    </div>
    <div
      class="styles_styles_bulletList_item__1am92eo2"
    >
      <span>
        root
      </span>
    </div>
    <div
      class="styles_styles_bulletList_child__1am92eo3"
    >
      <div
        class="styles_styles_bulletList_container__1am92eo0"
      >
        <div
          class="styles_styles_bulletList_bullet__1am92eo1"
        >
          •
        </div>
        <div
          class="styles_styles_bulletList_item__1am92eo2"
        >
          <span>
            foo
          </span>
        </div>
        <div
          class="styles_styles_bulletList_child__1am92eo3"
        />
      </div>
      <div
        class="styles_styles_bulletList_container__1am92eo0"
      >
        <div
          class="styles_styles_bulletList_bullet__1am92eo1"
        >
          •
        </div>
        <div
          class="styles_styles_bulletList_item__1am92eo2"
        >
          <span>
            bar
          </span>
        </div>
        <div
          class="styles_styles_bulletList_child__1am92eo3"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`add child 2`] = `
<div>
  <div
    class="styles_styles_bulletList_container__1am92eo0"
  >
    <div
      class="styles_styles_bulletList_bullet__1am92eo1"
    >
      •
    </div>
    <div
      class="styles_styles_bulletList_item__1am92eo2"
    >
      <span>
        root
      </span>
    </div>
    <div
      class="styles_styles_bulletList_child__1am92eo3"
    >
      <div
        class="styles_styles_bulletList_container__1am92eo0"
      >
        <div
          class="styles_styles_bulletList_bullet__1am92eo1"
        >
          •
        </div>
        <div
          class="styles_styles_bulletList_item__1am92eo2"
        >
          <span>
            foo
          </span>
        </div>
        <div
          class="styles_styles_bulletList_child__1am92eo3"
        />
      </div>
      <div
        class="styles_styles_bulletList_container__1am92eo0"
      >
        <div
          class="styles_styles_bulletList_bullet__1am92eo1"
        >
          •
        </div>
        <div
          class="styles_styles_bulletList_item__1am92eo2"
        >
          <span>
            bar
          </span>
        </div>
        <div
          class="styles_styles_bulletList_child__1am92eo3"
        />
      </div>
      <div
        class="styles_styles_bulletList_container__1am92eo0"
      >
        <div
          class="styles_styles_bulletList_bullet__1am92eo1"
        >
          •
        </div>
        <div
          class="styles_styles_bulletList_item__1am92eo2"
        >
          <span>
            baz
          </span>
        </div>
        <div
          class="styles_styles_bulletList_child__1am92eo3"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`inconsistent error > an item is not in linked list > error 1`] = `
<div>
  Error: there is an inconsistency in listItem in the children of '1': some listItems are no in linked list
</div>
`;

exports[`inconsistent error > an item is not in linked list > initial 1`] = `
<div>
  <div
    class="styles_styles_bulletList_container__1am92eo0"
  >
    <div
      class="styles_styles_bulletList_bullet__1am92eo1"
    >
      •
    </div>
    <div
      class="styles_styles_bulletList_item__1am92eo2"
    >
      <span>
        root
      </span>
    </div>
    <div
      class="styles_styles_bulletList_child__1am92eo3"
    >
      <div
        class="styles_styles_bulletList_container__1am92eo0"
      >
        <div
          class="styles_styles_bulletList_bullet__1am92eo1"
        >
          •
        </div>
        <div
          class="styles_styles_bulletList_item__1am92eo2"
        >
          <span>
            foo
          </span>
        </div>
        <div
          class="styles_styles_bulletList_child__1am92eo3"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`inconsistent error > first item not found > error 1`] = `
<div>
  Error: there is an inconsistency in listItem in the children of '1': no listItem with prevId = ''
</div>
`;

exports[`inconsistent error > first item not found > initial 1`] = `
<div>
  <div
    class="styles_styles_bulletList_container__1am92eo0"
  >
    <div
      class="styles_styles_bulletList_bullet__1am92eo1"
    >
      •
    </div>
    <div
      class="styles_styles_bulletList_item__1am92eo2"
    >
      <span>
        root
      </span>
    </div>
    <div
      class="styles_styles_bulletList_child__1am92eo3"
    >
      <div
        class="styles_styles_bulletList_container__1am92eo0"
      >
        <div
          class="styles_styles_bulletList_bullet__1am92eo1"
        >
          •
        </div>
        <div
          class="styles_styles_bulletList_item__1am92eo2"
        >
          <span>
            foo
          </span>
        </div>
        <div
          class="styles_styles_bulletList_child__1am92eo3"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`inconsistent error > next item not found > error 1`] = `
<div>
  Error: there is an inconsistency in listItem in the children of '1': no listItem with id = '3'
</div>
`;

exports[`inconsistent error > next item not found > initial 1`] = `
<div>
  <div
    class="styles_styles_bulletList_container__1am92eo0"
  >
    <div
      class="styles_styles_bulletList_bullet__1am92eo1"
    >
      •
    </div>
    <div
      class="styles_styles_bulletList_item__1am92eo2"
    >
      <span>
        root
      </span>
    </div>
    <div
      class="styles_styles_bulletList_child__1am92eo3"
    >
      <div
        class="styles_styles_bulletList_container__1am92eo0"
      >
        <div
          class="styles_styles_bulletList_bullet__1am92eo1"
        >
          •
        </div>
        <div
          class="styles_styles_bulletList_item__1am92eo2"
        >
          <span>
            foo
          </span>
        </div>
        <div
          class="styles_styles_bulletList_child__1am92eo3"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`not updated when locked > initial 1`] = `
<div>
  <div
    class="styles_styles_bulletList_container__1am92eo0"
  >
    <div
      class="styles_styles_bulletList_bullet__1am92eo1"
    >
      •
    </div>
    <div
      class="styles_styles_bulletList_item__1am92eo2"
    />
    <div
      class="styles_styles_bulletList_child__1am92eo3"
    />
  </div>
</div>
`;

exports[`not updated when locked > unlocked 1`] = `
<div>
  <div
    class="styles_styles_bulletList_container__1am92eo0"
  >
    <div
      class="styles_styles_bulletList_bullet__1am92eo1"
    >
      •
    </div>
    <div
      class="styles_styles_bulletList_item__1am92eo2"
    >
      <span>
        root
      </span>
    </div>
    <div
      class="styles_styles_bulletList_child__1am92eo3"
    >
      <div
        class="styles_styles_bulletList_container__1am92eo0"
      >
        <div
          class="styles_styles_bulletList_bullet__1am92eo1"
        >
          •
        </div>
        <div
          class="styles_styles_bulletList_item__1am92eo2"
        >
          <span>
            foo
          </span>
        </div>
        <div
          class="styles_styles_bulletList_child__1am92eo3"
        />
      </div>
      <div
        class="styles_styles_bulletList_container__1am92eo0"
      >
        <div
          class="styles_styles_bulletList_bullet__1am92eo1"
        >
          •
        </div>
        <div
          class="styles_styles_bulletList_item__1am92eo2"
        >
          <span>
            bar
          </span>
        </div>
        <div
          class="styles_styles_bulletList_child__1am92eo3"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`not updated when locked > updated with lock 1`] = `
<div>
  <div
    class="styles_styles_bulletList_container__1am92eo0"
  >
    <div
      class="styles_styles_bulletList_bullet__1am92eo1"
    >
      •
    </div>
    <div
      class="styles_styles_bulletList_item__1am92eo2"
    >
      <span>
        *root
      </span>
    </div>
    <div
      class="styles_styles_bulletList_child__1am92eo3"
    >
      <div
        class="styles_styles_bulletList_container__1am92eo0"
      >
        <div
          class="styles_styles_bulletList_bullet__1am92eo1"
        >
          •
        </div>
        <div
          class="styles_styles_bulletList_item__1am92eo2"
        >
          <span>
            *bar
          </span>
        </div>
        <div
          class="styles_styles_bulletList_child__1am92eo3"
        />
      </div>
      <div
        class="styles_styles_bulletList_container__1am92eo0"
      >
        <div
          class="styles_styles_bulletList_bullet__1am92eo1"
        >
          •
        </div>
        <div
          class="styles_styles_bulletList_item__1am92eo2"
        >
          <span>
            *foo
          </span>
        </div>
        <div
          class="styles_styles_bulletList_child__1am92eo3"
        />
      </div>
    </div>
  </div>
</div>
`;
